tools/utils.py

```diff
@@ -263,3 +266,95 @@ def normalize_image(
     return image_norm
+
+
+class BoundaryDataset(Dataset):
+    def __init__(self,
+                 img_paths: List[str],
+                 input_size: Union[int, List[int]] = (1024, 1024)):
+        self.img_paths = img_paths
+        self.input_size = (input_size, input_size) if isinstance(input_size, int) else input_size
+
+    def __len__(self):
+        return len(self.img_paths)
+
+    def __getitem__(self, idx):
+        img_path = self.img_paths[idx]
+        img = cv2.imread(img_path)
+        img = cv2.resize(img, self.input_size)
+        return img, img_path
+
+
+class MorphologicalTransformations:
+    def __init__(
+            self,
+            image_file: str,
+    ):
+
+        self.image_file = image_file
+        self.MAX_PIXEL = 255
+        self.MIN_PIXEL = 0
+
+    def read_image(self):
+        image_src = cv2.imread(self.image_file, 0)
+        return image_src
+
+    def binary(
+            self,
+            thresh_method: str = 'otsu',
+            thresh_val: float = None,
+    ):
+        color_1 = self.MAX_PIXEL
+        color_2 = self.MIN_PIXEL
+        image_src = self.read_image()
+
+        assert thresh_method in ['otsu', 'triangle', 'manual'], f'Invalid thresh_method: {thresh_method}'
+
+        if thresh_method == 'otsu':
+            threshold_value, _ = cv2.threshold(image_src, 0, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)
+        elif thresh_method == 'triangle':
+            threshold_value, _ = cv2.threshold(image_src, 0, 255, cv2.THRESH_BINARY + cv2.THRESH_TRIANGLE)
+        elif thresh_method == 'manual':
+            threshold_value = thresh_val
+        else:
+            logging.warning(f'Invalid threshold')
+
+        initial_conv = np.where((image_src <= threshold_value), image_src, color_1)
+        final_conv = np.where((initial_conv > threshold_value), initial_conv, color_2)
+
+        return final_conv
+
+    def erosion(
+            self,
+            image_src,
+    ):
+        kernel = np.ones((6, 6), 'uint8')
+        erode_img = cv2.erode(image_src, kernel, cv2.BORDER_REFLECT, iterations=1)
+        return erode_img
+
+    def extract_boundary(
+            self,
+            image_src,
+    ):
+        image_eroded = self.erosion(image_src=image_src)
+        ext_bound = image_src - image_eroded
+        return ext_bound
+
+    @staticmethod
+    def visualize_boundary(
+            image_src,
+            boundary,
+    ):  # boundary is the output from extract_boundary function
+        alpha = 0.5
+        beta = 1.0 - alpha
+        boundary = np.expand_dims(boundary, axis=-1)
+        res_bound = cv2.resize(boundary, dsize=(1024, 1024), interpolation=cv2.INTER_CUBIC)
+
+        image = Image.open(image_src)
+        image = image.resize((1024, 1024), Image.ANTIALIAS)
+        lung = np.expand_dims(image, axis=-1)
+        dst = cv2.addWeighted(res_bound, alpha, lung, beta, 0.0, dtype=cv2.CV_64F)
+        backtorgb = cv2.cvtColor(dst.astype(np.uint8), cv2.COLOR_GRAY2RGB)
```

Comment:
@sriva-e 

Why are we using two image processing libraries here: opencv and PIL? I would recommend using one of them, for instance OpenCV. It can do both reading images and resizing them afterward.